### DISCUSSION TITLE
Cacheable/shared props between pages on SSG

### DISCUSSION THREAD

**Discussion by aralroca** [vercel/next.js#19050] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

There are some cases, where several pages need the same JSON depending on the language or slugs.

Right now, if the pages `/en`, `/en/about`, `/en/contact` and some others share the same JSON as a prop. And this JSON is huge, navigating from `/en` to `/en/about` is downloading the same huge JSON object again, and the same navigating again from `/en/about` to `/en/contact`. This is because is downloading all the props of the page under `pagename.json`. Example navigating to `/` (`index.json`):

![image](https://user-images.githubusercontent.com/13313058/98802670-e632ce80-2413-11eb-8682-a7b78113014c.png)

In the first navigation, pages will always download their props, then depending on the browser will return a 304 and will not be so much problem. But this can affect the navigation between pages quite a bit if they use a recurrent prop.

## Describe the solution you'd like

Introducing cacheable props. 

```jsx
export default function ExamplePage({ common }) {
   return JSON.stringify(common)
}


export async function getStaticProps({ locale }) {
  return {
     cacheableProps: {
        common: { key: `common-${locale}`, value:  await getJSONFromDB(locale) }
     }
  }
}
```

Then, instead of when you navigate to another page it will download only `pagename.json`, it would also download the file `common-en.json`. The beauty of this is that by navigating to another page that has the same cacheableProp, then the browser will already return the file under **status 304**, so it will only download the `anotherpagename.json` file with the rest of props.

## Describe alternatives you've considered

Sure there are other ways, like to provide a different loader than getStaticProps or getServerSideProps, but I think the best alternative is to integrated into the existing ones.

## Additional context

Now that i18n is supported, it is a problem that I have found loading the translation dictionary that the pages have in common. Likewise, I'm sure this feature could be used for more things beyond i18n.

**Comment by aralroca** [comment]
Thinking about it, maybe the `cacheableProps` name is not the right one, maybe it would make more sense to use `sharedProps` or look for another one that fits better.

**Comment by aralroca** [comment]
Another alternative to declare this:

```js
export async function getStaticProps({ locale }) {
  return {
     props: {
        common: await getJSONFromDB(locale)
     },
     propsKeys: {
        common: `common-${locale}`
     }
  }
}
```

**Comment by awlevin** [comment]
Amazing. Came to this forum for this exact problem with a different use-case— glad to see I'm not alone.

Some NoSQL providers charge per document-read, which can add up quickly if you have thousands of pages with shared data. We could probably use Redis but that feels overkill for our situation. Since the shared data is of a manageable size, we periodically update it in a JSON file, which is less than ideal, but gets the job done.

Doing this manually is getting old though— does anyone know of any other approaches to streamline / automate this (even if they're hacky)? I love the thought of `cacheableProps` proposed in this thread but would be great to have another solution to hold us over in the meantime.

**Reply by elibenton** [reply]
I've had a similar use case, where calling an external API hits rate limits on every build. I've tried using a memory cache, but I can't get it to work.

Anyone found success?

**Reply by stigkj** [reply]
@elibenton you could use disk as a cache, but then the cache would be per instance. I guess the memory cache you were trying to do also would be per instance. Another option would be to use a Redis cluster as a cache.

**Reply by stigkj** [reply]
@awlevin Using a disk cache is very simple, although you would get a request per instance of the app.

**Reply by lukeshumard** [reply]
I've used [Cloudflare Workers](https://developers.cloudflare.com/workers/) for this and it works really well. You can either cache the response from your origin or use [Cloudflare's Key Value](https://developers.cloudflare.com/workers/learning/how-kv-works) for a more custom solution. You can also use Workers to create a URL that your external API / CMS can call when your content has changed (such as a webhook), and then invalidate your cache or update the Key Value. I found this much easier than using a disk cache.

**Comment by stigkj** [comment]
@aralroca you could use client-side fetch for either the common part or the specific part of the props and `getStaticProps` for the other part of the props.
